### ISSUE TITLE
multiple issues with numpy.f2py.compile() function

### ISSUE THREAD

**Issue by mbdevpl** [numpy/numpy#7683]
I've encountered multiple problems when using `numpy.f2py.compile()` function on Windows and Ubuntu. For now, the issues are with Python 3.5.1, Numpy 1.11.0 on Windows 7 x64 and Ubuntu 14.04 x64

1. docstring of `compile()` is completely missing the help for parameter `source_fn`. This is certainly platform independent issue...

2. when running `compile()` the following messages show up:

On Windows:

> Could not locate executable C:ProgramsPython35python.exe
> Executable C:ProgramsPython35python.exe does not exist

> C:\Programs\Python35\lib\site-packages\numpy\distutils\misc_util.py:6: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
> import imp

On Ubuntu:

> /home/mb/.local/lib/python3.5/site-packages/numpy/distutils/misc_util.py:6: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp

**Comment by mbdevpl** [comment]
I took a look into the source code where those errors come from, and the code is pretty... old. Do you think it's doable for someone with little knowledge about NumPy's source code, but overall experience with Python to go ahead and fix those, or did I stumble upon a kind of dark corner of NumPy where no-one should look? ;)

**Comment by rgommers** [comment]
``f2py`` and ``distutils`` are two of the darker corners of numpy indeed:) But in this case it doesn't seem too difficult to fix. The docstring edit is the only thing needed in ``f2py`` itself. So the main thing is to figure out exactly how to change ``imp._load_module`` usage into ``importlib.exec_module`` for Python >= 3.4

**Comment by bertrand-l** [comment]
@mbdevpl are you working on this? looked into this as i was curious about the "dark corners", and i might have a fix

**Comment by mbdevpl** [comment]
Not working on it right now, so it's great if you'd have a fix. I looked through your commit and it seems OK fix for the warning:

> C:\Programs\Python35\lib\site-packages\numpy\distutils\misc_util.py:6: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
> import imp

though obviously someone from NumPy team should comment :)

On the other hand, still, issue of docstring of `compile()` and below strange output remains:

> Could not locate executable C:ProgramsPython35python.exe
> Executable C:ProgramsPython35python.exe does not exist

I wonder if I should create separate issues for those.

**Comment by bertrand-l** [comment]
Ok, thanks. I have updated the `compile` docstring in a separate commit. Not sure what to think of the `C:ProgramsPython35python.exe` issue but I will have look (probably only tomorrow though)

**Comment by charris** [comment]
Can this be closed? The `Could not locate executable C:ProgramsPython35python.exe` seems the only thing remaining and that seems to be unrelated to numpy.

**Comment by bertrand-l** [comment]
I could not find anything in the numpy code that could explain `Could not locate executable C:ProgramsPython35python.exe`. If this persists, I would suggest the OP to open a new issue, indicating what versions of windows, Python and numpy are used.

**Comment by mbdevpl** [comment]
Yes, I also couldn't find anything in the source that explains the `Could not locate...` message. It seems Windows-specific, because this kinds of messages show up when you attempt to launch a non-existing program using `cmd.exe`. The path `C:\Programs\Python35\python.exe` is where my Python executable is located. The message shows up while `f2py.compile()` is running, so I know it's somewhere inside there... It looks like some part of f2py or tool used by it is trying to launch python executable without backslashes between dirnames, but I didn't find where specifically it happens. If I ever learn anything more about it I'll post a new issue, you can close this issue at will.

Thanks @bertrand-l and @charris for help with all this!

**Comment by davidjgall** [comment]
I think the message is triggered when no fortran compiler is found. The python executable is actually found, despite the screwed up path/executable that gets reported. The concatenated path/filename seems to be a bit of Windows (DOS) code that's losing the backslashes in the process of generating the text that becomes the error message, but the message itself is actually only triggered by a "failure" result code returned by the "missing" executable (which was actually found). The actual failure is in not finding a gfortran.exe or other fortran compiler on the path, which causes a python.exe "failure" result code which, in turn, triggers the Windows (DOS) message containing the garbled textual rendering of the path/filename to the python executable.
